### ISSUE TITLE
Parallelizing cross validation

### ISSUE THREAD

**Issue by ckemere** [dask/dask#2804]
I have a nested loop that I'm trying to parallelize that looks like this
````
def est_model(data, params):
   return model

def score_data(data, model):
   return score

for dataset in Datasets:
    raw_obj = BigData[dataset] ## WANT TO WRAP THIS IN DELAYED
    processed_obj = raw_obj.process() ## TO DELAY THIS
   
    X = list(range(processed_obj.n_chunks))
    scores = []
    kf = KFold(n_splits=5)

    for training, validation in kf.split(X):     
        train = processed_obj[training]
        test = processed_obj[validation]
        model = delayed(est_model)(delayed(train, traverse=False), params)
        scores.append(delayed(score_data)(delayed(test, traverse=False), model))
    
    scores_by_dataset.append(scores)

sbs = dask.compute(*scores_by_datsaset)
````

It currently works fine, but I'd like to further parallelize the initial raw data processing. The problem is that knowing the number of chunks requires this pre-processing step. As far as I can see, this will mean that there is a for loop which depends on a delayed object. Is there a pattern for cross-validation that I'm just missing?

**Comment by mrocklin** [comment]
Are you already familiar with http://dask-searchcv.readthedocs.io/en/latest/ ?

**Comment by ckemere** [comment]
Unfortunately, the broader analysis scenario is model comparison rather than parameter optimization. (So we run the cross-validated scoring across different data sets and compare the results.)

**Comment by kdubovikov** [comment]
Interested on how we can do simple CV without parameter search with dask too

**Comment by mrocklin** [comment]
What are common ways to split a dataset?  I assume that your `for dataset in Datasets` Datasets list is a list of slices.  Are these random indices?  If so and if you ever pass through dataframes then `df.random_split` might be of use.  http://dask.pydata.org/en/latest/dataframe-api.html#dask.dataframe.Series.random_split .  If you don't pass through dataframes then you might already know the shape of your array, and so could presumably use slices or the `da.random` module.

Can you say more about the common constraints on cross validation?

Alternatively, @TomAugspurger and @ogrisel probably know more than I do here.

**Comment by kdubovikov** [comment]
@mrocklin, thanks for the response. The general use case is:

1. Split a dataframe (preferably with a stratified split, so not random) to n folds. Each fold is a split by a fraction of rows x (x < 1).
2. For each split 1..n: (map)
2.1. Fit a model on each cluster node in parallel using first part on the split 
2.2. Evaluate model using some kind of metric like RMSE, AUC or cross-entropy
3. Collect results for each fold and average them (reduce)
4. Select the best model according to average metric value from 2.2

**Comment by TomAugspurger** [comment]
Thanks @kdubovikov 

I'm hoping to implement some CV strategies (and utilities like cross_val_score) in dask-ml this week. I'll update this issue when that's done.

**Comment by yipcma** [comment]
@TomAugspurger  some head-way on simple cross_val_score?

**Comment by TomAugspurger** [comment]
https://github.com/dask/dask-ml/compare/master...TomAugspurger:train-test-split?expand=1 has `train_test_split` mostly finished I think. I don't recall why I didn't make a PR with it yet.

Haven't had time to work on the rest, so if anyone is interested in finishing off that PR or picking up the follow-on things like `cross_val_score` and the various CV strategies I'd appreciate it.

**Comment by BrendanMartin** [comment]
@TomAugspurger has that been merged yet?

**Comment by TomAugspurger** [comment]
@BrendanMartin unfortunately no one has had time to finish it off. Hopefully soon, though perhaps you'd be interested in finishing it off before I get around to it? ;)

See also https://github.com/dask/dask-ml/issues/165

**Comment by BrendanMartin** [comment]
I wish I could finish it, but unfortunately I'm not super familiar with how it all works and what even needs to be done.

**Comment by morganics** [comment]
Does this seem like a reasonable way of doing a shuffled KFold on a dask dataframe?

```
def exclude(lst, i):
    if i == 0:
        return lst[i+1:]
    return lst[:i] + lst[i+1:]

def split(self, ddf, n_splits):
    partitions = ddf.random_split((1 / n_splits) * np.ones(n_splits, dtype=np.int))
    for split in range(n_splits):
        training_partitions = exclude(partitions, split)
        yield (dd.concat(training_partitions, axis=0, interleave_partitions=True), partitions[split])

for (train, test) in split(ddf, 4):
    ...
```

**Comment by mrocklin** [comment]
Curious, what's the status on this?

**Comment by TomAugspurger** [comment]
dask-ml implemented train_test_split, ShuffleSplit, and KFold cross validators. If people want additional CV routines (say stratified) then I'd recommend opening an [issue on dask-ml](https://github.com/dask/dask-ml/issues/new) so we can prioritize that.

**Comment by EmanuelFontelles** [comment]
How can I train a GIANT dataset with Dask? Have I to split the data and train separately? How can I measure accuracy in this case?

**Comment by jrbourbeau** [comment]
@EmanuelFontelles I'd recommend checking out dask-ml (http://ml.dask.org/). It has estimators, metrics like `accuracy_score`, etc. that work with dask arrays and dask DataFrames.
